File: apps/api/src/routes/preview.ts
import { createHash } from "node:crypto";
import type { FastifyInstance } from "fastify";
import { pool } from "../db/pool.js";
import { requireAuth } from "../middleware/auth.js";
import {
  resolveEffectivePermission,
  type DocumentRow,
} from "../services/document-permissions.js";
import { resolveEditorRoute } from "../services/editor-types.js";
import {
  createObjectStorage,
  objectKeyForVersion,
} from "../services/object-storage.js";
import { config } from "../config.js";

const storage = createObjectStorage();

function minimalPdfBuffer(title: string): Buffer {
  const text = `OFD Preview: ${title}`;
  const pdf = `%PDF-1.4
1 0 obj<</Type/Catalog/Pages 2 0 R>>endobj
2 0 obj<</Type/Pages/Kids[3 0 R]/Count 1>>endobj
3 0 obj<</Type/Page/MediaBox[0 0 612 792]/Parent 2 0 R/Contents 4 0 R>>endobj
4 0 obj<</Length ${text.length + 20}>>stream
BT /F1 12 Tf 72 720 Td (${text}) Tj ET
endstream endobj
xref
0 5
trailer<</Size 5/Root 1 0 R>>
startxref
0
%%EOF`;
  return Buffer.from(pdf, "utf8");
}

export async function registerPreviewRoutes(app: FastifyInstance) {
  app.get("/api/preview/:documentId", async (request, reply) => {
    const user = requireAuth(request);
    const { documentId } = request.params as { documentId: string };

    const client = await pool.connect();
    try {
      const docRes = await client.query(
        `SELECT d.*, dv.version_id, dv.object_key, dv.file_hash
         FROM documents d
         JOIN document_versions dv ON d.current_version_id = dv.version_id
         WHERE d.document_id = $1 AND d.is_deleted = FALSE`,
        [documentId],
      );
      if (!docRes.rows.length) {
        return reply.status(404).send({ error: "文档不存在" });
      }

      const row = docRes.rows[0];
      const doc = row as DocumentRow & {
        version_id: string;
        object_key: string;
        file_hash: string;
        mime_type?: string;
      };

      if (doc.tenant_id !== user.tenantId) {
        return reply.status(403).send({ error: "无权限" });
      }

      const level = await resolveEffectivePermission(client, user, doc);
      if (level === "none") {
        return reply.status(403).send({ error: "无权限" });
      }

      const routeInfo = resolveEditorRoute(doc.name);

      if (routeInfo.route === "preview-ofd") {
        const cacheKey = createHash("sha256")
          .update(`${doc.tenant_id}:${doc.file_hash}`)
          .digest("hex");
        const cacheRes = await client.query(
          `SELECT target_object_key FROM editor_conversion_cache WHERE source_hash = $1`,
          [cacheKey],
        );

        let previewKey: string;
        if (cacheRes.rows.length) {
          previewKey = cacheRes.rows[0].target_object_key as string;
        } else {
          const pdfBuf = minimalPdfBuffer(doc.name);
          previewKey = objectKeyForVersion(
            doc.tenant_id,
            documentId,
            `ofd-preview-${cacheKey.slice(0, 8)}`,
          );
          await storage.put(previewKey, pdfBuf, "application/pdf");
          await client.query(
            `INSERT INTO editor_conversion_cache (source_hash, target_object_key, target_mime)
             VALUES ($1, $2, 'application/pdf')
             ON CONFLICT (source_hash) DO NOTHING`,
            [cacheKey, previewKey],
          );
        }

        const url = await storage.presignedUrl(previewKey, 300);
        return {
          previewType: "ofd",
          label: "只读预览（OFD 转 PDF）",
          readOnly: true,
          url,
          dsUrl: config.onlyoffice.dsUrl,
          aiEntries: ["aimed", "translation"],
        };
      }

      if (routeInfo.route === "preview-pdf") {
        const url = await storage.presignedUrl(doc.object_key, 300);
        return {
          previewType: "pdf",
          readOnly: true,
          url,
          dsUrl: config.onlyoffice.dsUrl,
          aiEntries: ["aimed", "translation"],
          currentPage: 1,
        };
      }

      if (routeInfo.route === "preview-image") {
        const url = await storage.presignedUrl(doc.object_key, 300);
        return {
          previewType: "image",
          url,
          fileHash: doc.file_hash,
          visualParse: true,
        };
      }

      return reply.status(400).send({ error: "不支持的预览类型" });
    } finally {
      client.release();
    }
  });

  app.get("/api/preview/:documentId/parse-status", async (request, reply) => {
    const user = requireAuth(request);
    const { documentId } = request.params as { documentId: string };

    const client = await pool.connect();
    try {
      const docRes = await client.query(
        `SELECT * FROM documents WHERE document_id = $1 AND tenant_id = $2`,
        [documentId, user.tenantId],
      );
      if (!docRes.rows.length) {
        return reply.status(404).send({ error: "文档不存在" });
      }

      // document_parse_jobs 由 c03 owner 建表（migration 005）；建表前优雅降级避免 42P01
      const reg = await client.query(
        `SELECT to_regclass('public.document_parse_jobs') AS t`,
      );
      if (!reg.rows[0]?.t) {
        return {
          status: "pending",
          message: "等待 c03 解析服务建表并消费 upload_success 事件后创建作业",
          jobs: [],
        };
      }

      // 1.5a：用 document_version + c03 状态词（pending/parsing/succeeded/failed），
      // JOIN document_visual_parse_results 取结构化结果；去除 stub 列 result/job_type/version_id
      const jobRes = await client.query(
        `SELECT j.job_id, j.status, j.substatus, j.failure_reason, j.document_version,
                j.index_ready_at, j.updated_at,
                r.confidence AS visual_confidence, r.failure_reason AS visual_failure_reason
         FROM document_parse_jobs j
         LEFT JOIN document_visual_parse_results r
           ON r.document_id = j.document_id AND r.document_version = j.document_version
         WHERE j.document_id = $1 AND j.tenant_id = $2
         ORDER BY j.created_at DESC LIMIT 1`,
        [documentId, user.tenantId],
      );

      if (!jobRes.rows.length) {
        return {
          status: "pending",
          message: "等待 c03 解析服务消费 upload_success 事件后创建作业",
          jobs: [],
        };
      }

      const row = jobRes.rows[0];
      return {
        status: row.status,
        substatus: row.substatus,
        documentVersion: row.document_version,
        failureReason: row.failure_reason,
        indexReadyAt: row.index_ready_at,
        updatedAt: row.updated_at,
        visual:
          row.visual_confidence !== null || row.visual_failure_reason !== null
            ? { confidence: row.visual_confidence, failureReason: row.visual_failure_reason }
            : null,
        jobs: jobRes.rows,
      };
    } finally {
      client.release();
    }
  });
}

File: apps/server/internal/routes/preview.go
package routes

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"net/http"
	"time"

	"github.com/gin-gonic/gin"
	"gorm.io/gorm"

	"medoffice/server/internal/auth"
	"medoffice/server/internal/config"
	"medoffice/server/internal/docperm"
	"medoffice/server/internal/editor"
	"medoffice/server/internal/httpx"
	"medoffice/server/internal/storage"
)

func minimalPdfBuffer(title string) []byte {
	text := "OFD Preview: " + title
	pdf := fmt.Sprintf(`%%PDF-1.4
1 0 obj<</Type/Catalog/Pages 2 0 R>>endobj
2 0 obj<</Type/Pages/Kids[3 0 R]/Count 1>>endobj
3 0 obj<</Type/Page/MediaBox[0 0 612 792]/Parent 2 0 R/Contents 4 0 R>>endobj
4 0 obj<</Length %d>>stream
BT /F1 12 Tf 72 720 Td (%s) Tj ET
endstream endobj
xref
0 5
trailer<</Size 5/Root 1 0 R>>
startxref
0
%%%%EOF`, len([]rune(text))+20, text)
	return []byte(pdf)
}

// RegisterPreview 复刻 routes/preview.ts：预览（pdf/image/ofd）+ parse-status（to_regclass 降级）。
func RegisterPreview(r *gin.Engine, db *gorm.DB, store *storage.Storage, cfg config.OnlyOffice) {
	r.GET("/api/preview/:documentId", func(c *gin.Context) {
		user, ok := auth.Require(c)
		if !ok {
			return
		}
		documentID := c.Param("documentId")
		var row struct {
			DocumentID string `gorm:"column:document_id"`
			TenantID   string `gorm:"column:tenant_id"`
			OwnerID    string `gorm:"column:owner_id"`
			Name       string `gorm:"column:name"`
			Space      string `gorm:"column:space"`
			ObjectKey  string `gorm:"column:object_key"`
			FileHash   string `gorm:"column:file_hash"`
		}
		_ = db.Raw(
			`SELECT d.*, dv.version_id, dv.object_key, dv.file_hash
			 FROM documents d JOIN document_versions dv ON d.current_version_id = dv.version_id
			 WHERE d.document_id = ? AND d.is_deleted = FALSE`, documentID,
		).Scan(&row).Error
		if row.DocumentID == "" {
			httpx.Fail(c, 404, "文档不存在")
			return
		}
		if row.TenantID != user.TenantID {
			httpx.Fail(c, 403, "无权限")
			return
		}
		doc := docperm.DocumentRow{DocumentID: row.DocumentID, TenantID: row.TenantID, OwnerID: row.OwnerID, Space: row.Space}
		level, _ := docperm.Resolve(db, user, doc)
		if level == docperm.None {
			httpx.Fail(c, 403, "无权限")
			return
		}
		info := editor.ResolveEditorRoute(row.Name)
		ctx := c.Request.Context()

		switch info.Route {
		case editor.RoutePreviewOFD:
			sum := sha256.Sum256([]byte(row.TenantID + ":" + row.FileHash))
			cacheKey := hex.EncodeToString(sum[:])
			var previewKey string
			_ = db.Raw(`SELECT target_object_key FROM editor_conversion_cache WHERE source_hash = ?`, cacheKey).Scan(&previewKey).Error
			if previewKey == "" {
				pdfBuf := minimalPdfBuffer(row.Name)
				previewKey = storage.ObjectKeyForVersion(row.TenantID, documentID, "ofd-preview-"+cacheKey[:8])
				if err := store.Put(ctx, previewKey, pdfBuf, "application/pdf"); err != nil {
					httpx.Fail(c, 500, "服务器错误")
					return
				}
				_ = db.Exec(
					`INSERT INTO editor_conversion_cache (source_hash, target_object_key, target_mime)
					 VALUES (?, ?, 'application/pdf') ON CONFLICT (source_hash) DO NOTHING`,
					cacheKey, previewKey,
				).Error
			}
			url, err := store.PresignedURL(ctx, previewKey, 300*time.Second)
			if err != nil {
				httpx.Fail(c, 500, "服务器错误")
				return
			}
			c.JSON(http.StatusOK, gin.H{
				"previewType": "ofd", "label": "只读预览（OFD 转 PDF）", "readOnly": true,
				"url": url, "dsUrl": cfg.DSURL, "aiEntries": []string{"aimed", "translation"},
			})
		case editor.RoutePreviewPDF:
			url, err := store.PresignedURL(ctx, row.ObjectKey, 300*time.Second)
			if err != nil {
				httpx.Fail(c, 500, "服务器错误")
				return
			}
			c.JSON(http.StatusOK, gin.H{
				"previewType": "pdf", "readOnly": true, "url": url, "dsUrl": cfg.DSURL,
				"aiEntries": []string{"aimed", "translation"}, "currentPage": 1,
			})
		case editor.RoutePreviewImage:
			url, err := store.PresignedURL(ctx, row.ObjectKey, 300*time.Second)
			if err != nil {
				httpx.Fail(c, 500, "服务器错误")
				return
			}
			c.JSON(http.StatusOK, gin.H{
				"previewType": "image", "url": url, "fileHash": row.FileHash, "visualParse": true,
			})
		default:
			httpx.Fail(c, 400, "不支持的预览类型")
		}
	})

	r.GET("/api/preview/:documentId/parse-status", func(c *gin.Context) {
		user, ok := auth.Require(c)
		if !ok {
			return
		}
		documentID := c.Param("documentId")
		var exists string
		_ = db.Raw(`SELECT document_id FROM documents WHERE document_id = ? AND tenant_id = ?`, documentID, user.TenantID).Scan(&exists).Error
		if exists == "" {
			httpx.Fail(c, 404, "文档不存在")
			return
		}

		// document_parse_jobs 由 c03 建表（migration 005）；建表前优雅降级避免 42P01
		var reg *string
		_ = db.Raw(`SELECT to_regclass('public.document_parse_jobs')::text AS t`).Scan(&reg).Error
		if reg == nil || *reg == "" {
			c.JSON(http.StatusOK, gin.H{
				"status": "pending", "jobs": []any{},
				"message": "等待 c03 解析服务建表并消费 upload_success 事件后创建作业",
			})
			return
		}

		var jobs []struct {
			JobID               string     `gorm:"column:job_id" json:"job_id"`
			Status              string     `gorm:"column:status" json:"status"`
			Substatus           *string    `gorm:"column:substatus" json:"substatus"`
			FailureReason       *string    `gorm:"column:failure_reason" json:"failure_reason"`
			DocumentVersion     int        `gorm:"column:document_version" json:"document_version"`
			IndexReadyAt        *time.Time `gorm:"column:index_ready_at" json:"index_ready_at"`
			UpdatedAt           time.Time  `gorm:"column:updated_at" json:"updated_at"`
			VisualConfidence    *float64   `gorm:"column:visual_confidence" json:"visual_confidence"`
			VisualFailureReason *string    `gorm:"column:visual_failure_reason" json:"visual_failure_reason"`
		}
		_ = db.Raw(
			`SELECT j.job_id, j.status, j.substatus, j.failure_reason, j.document_version,
			        j.index_ready_at, j.updated_at,
			        r.confidence AS visual_confidence, r.failure_reason AS visual_failure_reason
			 FROM document_parse_jobs j
			 LEFT JOIN document_visual_parse_results r
			   ON r.document_id = j.document_id AND r.document_version = j.document_version
			 WHERE j.document_id = ? AND j.tenant_id = ?
			 ORDER BY j.created_at DESC LIMIT 1`, documentID, user.TenantID,
		).Scan(&jobs).Error
		if len(jobs) == 0 {
			c.JSON(http.StatusOK, gin.H{
				"status": "pending", "jobs": []any{},
				"message": "等待 c03 解析服务消费 upload_success 事件后创建作业",
			})
			return
		}
		j := jobs[0]
		var visual any
		if j.VisualConfidence != nil || j.VisualFailureReason != nil {
			visual = gin.H{"confidence": j.VisualConfidence, "failureReason": j.VisualFailureReason}
		}
		c.JSON(http.StatusOK, gin.H{
			"status": j.Status, "substatus": j.Substatus, "documentVersion": j.DocumentVersion,
			"failureReason": j.FailureReason, "indexReadyAt": j.IndexReadyAt, "updatedAt": j.UpdatedAt,
			"visual": visual, "jobs": jobs,
		})
	})
}
